File: hospital-ai/frontend/src/types/scheduling.ts
export interface DoctorCandidate {
  doctor_id: string;
  doctor_name: string;
  specialization: string;
  department_id: string | null;
  department_name: string | null;
  availability_status: string;
  workload_count: number;
  score: number;
  reasons: string[];
}

export interface DoctorAssignmentResult {
  selected_doctor_id: string | null;
  selected_doctor_name: string | null;
  selection_score: number;
  candidates: DoctorCandidate[];
  rationale: string[];
}

export interface SlotRecommendation {
  doctor_id: string;
  doctor_name: string;
  appointment_date: string;
  start_time: string;
  end_time: string;
  score: number;
  reasons: string[];
}

export interface AppointmentSchedulingResult {
  recommended_slot: SlotRecommendation | null;
  alternatives: SlotRecommendation[];
  booking_status: string;
  rationale: string[];
}

export interface SurgerySchedulingResult {
  required: boolean;
  procedure_names: string[];
  duration_minutes: number | null;
  recommended_slot: SlotRecommendation | null;
  operation_theatre_status: string;
  resource_allocation_required: boolean;
  notes: string[];
}

export interface FollowUpPlan {
  recommended_date: string | null;
  interval_days: number | null;
  reason: string;
  planning_only: boolean;
}

export interface QueueItem {
  appointment_id: string;
  patient_id: string;
  patient_name: string;
  start_time: string;
  end_time: string;
  priority_level: string;
  priority_score: number;
  position: number;
}

export interface QueueOptimizationResult {
  doctor_id: string | null;
  appointment_date: string | null;
  ordered_queue: QueueItem[];
  changed_order_count: number;
  rationale: string[];
}

export interface DoctorWorkload {
  doctor_id: string;
  doctor_name: string;
  active_appointments: number;
  workload_score: number;
  availability_status: string;
}

export interface WorkloadBalancingResult {
  horizon_days: number;
  doctor_loads: DoctorWorkload[];
  balance_gap: number;
  recommendation: string;
}

export interface SchedulingResult {
  id: string;
  patient_id: string;
  processing_job_id: string | null;
  doctor_assignment_json: DoctorAssignmentResult;
  appointment_scheduling_json: AppointmentSchedulingResult;
  surgery_scheduling_json: SurgerySchedulingResult;
  follow_up_planning_json: FollowUpPlan;
  queue_optimization_json: QueueOptimizationResult;
  workload_balancing_json: WorkloadBalancingResult;
  summary: string | null;
  engine: string;
  emergency_priority_level: string;
  emergency_priority_score: number;
  visit_type: string;
  derived_department: string | null;
  derived_specialists_json: string[];
  procedures_json: string[];
  surgery_recommendation_json: string[];
  surgery_sources_json: Record<string, boolean>;
  surgery_conflict: boolean;
  source_result_ids_json: Record<string, string | null>;
  source_availability_json: Record<string, boolean>;
  recommended_tests_json: string[];
  recommended_imaging_json: string[];
  recommended_medications_json: string[];
  treatment_modes_json: string[];
  resource_requirements_json: string[];
  treatment_validation_status: string | null;
  status: string;
  warnings_json: string[];
  processing_time_ms: number | null;
  created_at: string;
  updated_at: string | null;
}

export interface SchedulingStartResult {
  patient_id: string;
  status: string;
  processing_time_ms: number;
  summary: string;
  engine: string;
  emergency_priority_level: string;
  emergency_priority_score: number;
  visit_type: string;
  derived_department: string | null;
  derived_specialists: string[];
  procedures: string[];
  surgery_recommendation: string[];
  surgery_sources: Record<string, boolean>;
  surgery_conflict: boolean;
  source_result_ids: Record<string, string | null>;
  source_availability: Record<string, boolean>;
  recommended_tests: string[];
  recommended_imaging: string[];
  recommended_medications: string[];
  treatment_modes: string[];
  resource_requirements: string[];
  treatment_validation_status: string | null;
  warnings: string[];
  doctor_assignment: DoctorAssignmentResult;
  appointment_scheduling: AppointmentSchedulingResult;
  surgery_scheduling: SurgerySchedulingResult;
  follow_up_planning: FollowUpPlan;
  queue_optimization: QueueOptimizationResult;
  workload_balancing: WorkloadBalancingResult;
  scheduling_result: SchedulingResult;
}

File: hospital-ai/backend/app/schemas/scheduling.py
"""Pydantic schemas for the Scheduling Agent API."""
from __future__ import annotations
from datetime import date
from typing import Any, Dict, List
from uuid import UUID
from pydantic import BaseModel, Field
from app.ai.scheduling.models import (
    DoctorAssignmentResult, AppointmentSchedulingResult, SurgerySchedulingResult,
    FollowUpPlan, QueueOptimizationResult, WorkloadBalancingResult,
)

class SchedulingStartRequest(BaseModel):
    patient_id: UUID
    preferred_date: date = Field(default_factory=date.today)

class SchedulingResultOut(BaseModel):
    id: UUID
    patient_id: UUID
    processing_job_id: UUID | None = None
    doctor_assignment_json: Dict[str,Any] = Field(default_factory=dict)
    appointment_scheduling_json: Dict[str,Any] = Field(default_factory=dict)
    surgery_scheduling_json: Dict[str,Any] = Field(default_factory=dict)
    follow_up_planning_json: Dict[str,Any] = Field(default_factory=dict)
    queue_optimization_json: Dict[str,Any] = Field(default_factory=dict)
    workload_balancing_json: Dict[str,Any] = Field(default_factory=dict)
    summary: str | None = None
    engine: str = "deterministic_scheduling_rules_v1"
    emergency_priority_level: str = "Routine"
    emergency_priority_score: float = 0.0
    visit_type: str = "Consultation"
    derived_department: str | None = None
    derived_specialists_json: List[str] = Field(default_factory=list)
    surgery_recommendation_json: List[str] = Field(default_factory=list)
    source_result_ids_json: Dict[str, str | None] = Field(default_factory=dict)
    source_availability_json: Dict[str, bool] = Field(default_factory=dict)
    recommended_tests_json: List[str] = Field(default_factory=list)
    recommended_imaging_json: List[str] = Field(default_factory=list)
    recommended_medications_json: List[str] = Field(default_factory=list)
    treatment_validation_status: str | None = None
    status: str = "Completed"
    warnings_json: List[str] = Field(default_factory=list)
    processing_time_ms: int | None = None
    created_at: Any
    updated_at: Any = None

class SchedulingStartResponse(BaseModel):
    patient_id: UUID
    status: str = "Completed"
    processing_time_ms: int
    summary: str
    engine: str
    emergency_priority_level: str = "Routine"
    emergency_priority_score: float = 0.0
    visit_type: str = "Consultation"
    derived_department: str | None = None
    derived_specialists: List[str] = Field(default_factory=list)
    surgery_recommendation: List[str] = Field(default_factory=list)
    source_result_ids: Dict[str, str | None] = Field(default_factory=dict)
    source_availability: Dict[str, bool] = Field(default_factory=dict)
    recommended_tests: List[str] = Field(default_factory=list)
    recommended_imaging: List[str] = Field(default_factory=list)
    recommended_medications: List[str] = Field(default_factory=list)
    treatment_validation_status: str | None = None
    warnings: List[str] = Field(default_factory=list)
    doctor_assignment: DoctorAssignmentResult
    appointment_scheduling: AppointmentSchedulingResult
    surgery_scheduling: SurgerySchedulingResult
    follow_up_planning: FollowUpPlan
    queue_optimization: QueueOptimizationResult
    workload_balancing: WorkloadBalancingResult
    scheduling_result: SchedulingResultOut
